src/Nethermind/Nethermind.Blockchain/BlockTree.cs

```diff
@@ -1897,22 +1903,29 @@ private bool ShouldCache(long number)
                 {
                     // TODO: this is here because storing block data is not transactional
                     // TODO: would be great to remove it, he?
-                    if (_logger.IsTrace)
-                        _logger.Trace(
-                            $"Entering missing block info in {nameof(FindBlock)} scope when head is {Head?.ToString(Block.Format.Short)}");
-                    if (calculateTotalDifficulty)
+                    if (_logger.IsWarn) _logger.Warn($"Entering missing block info in {nameof(FindBlock)} scope when head is {Head?.ToString(Block.Format.Short)}. BlockHeader {block.ToString(Block.Format.FullHashAndNumber)}, CreateLevelIfMissing: {createLevelIfMissing}. BestKnownBeaconNumber: {BestKnownBeaconNumber}, BestKnownNumber: {BestKnownNumber}");
+                    if (createLevelIfMissing == false || (BestKnownBeaconNumber > BestKnownNumber && block.Number > BestKnownNumber))
+                    {
+                        if (_logger.IsWarn) _logger.Warn($"Ignoring creation of the level in {nameof(FindBlock)} scope when head is {Head?.ToString(Block.Format.Short)}. BlockHeader {block.ToString(Block.Format.FullHashAndNumber)}, CreateLevelIfMissing: {createLevelIfMissing}. BestKnownBeaconNumber: {BestKnownBeaconNumber}, BestKnownNumber: {BestKnownNumber}");
+                    }
+                    else
                     {
                         SetTotalDifficulty(block.Header);
+                        blockInfo = new BlockInfo(block.Hash, block.TotalDifficulty ?? UInt256.Zero);
+                        level = UpdateOrCreateLevel(block.Number, block.Hash, blockInfo);
                     }
-
-                    blockInfo = new BlockInfo(block.Hash, block.TotalDifficulty ?? UInt256.Zero);
-                    level = UpdateOrCreateLevel(block.Number, block.Hash, blockInfo, false);
                 }
                 else
                 {
-                    block.Header.TotalDifficulty = blockInfo.TotalDifficulty;
+                    if (blockInfo.TotalDifficulty != 0 || block.IsGenesis)
+                        block.Header.TotalDifficulty = blockInfo.TotalDifficulty;
+                    else
+                    {
+                        _logger.Info("Skipped...");
+                    }
                 }
 
+
```

Comment:
Same as above, is it copy-paste?

Reply:
FindHeader and FindBlock were similar before